src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs

```diff
@@ -2,45 +2,101 @@
-            // we do a lazy initialization here to make .NET Native builds happy.
-            if (_loadContext == null)
+            if (!ShouldLoad(fullPath))
             {
-                AssemblyLoadContext loadContext = AssemblyLoadContext.GetLoadContext(typeof(DefaultAnalyzerAssemblyLoader).GetTypeInfo().Assembly);
+                throw new ArgumentException(message: null, paramName: nameof(fullPath));
+            }
+
+            AssemblyLoadContext? loadContext;
 
-                if (System.Threading.Interlocked.CompareExchange(ref _loadContext, loadContext, null) == null)
+            var fullDirectoryPath = Path.GetDirectoryName(fullPath) ?? throw new ArgumentException(message: null, paramName: nameof(fullPath));
+            lock (_guard)
+            {
+                if (!_loadContextByDirectory.TryGetValue(fullDirectoryPath, out loadContext))
                 {
-                    _loadContext.Resolving += (context, name) =>
-                    {
-                        Debug.Assert(ReferenceEquals(context, _loadContext));
-                        return Load(name.FullName);
-                    };
+                    loadContext = new DirectoryLoadContext(fullDirectoryPath, this);
+                    _loadContextByDirectory[fullDirectoryPath] = loadContext;
                 }
             }
 
-            return LoadImpl(fullPath);
+            return loadContext.LoadFromAssemblyPath(fullPath);
         }
 
-        protected virtual Assembly LoadImpl(string fullPath) => _loadContext.LoadFromAssemblyPath(fullPath);
+        private sealed class DirectoryLoadContext : AssemblyLoadContext
+        {
+            private readonly string _directory;
+            private readonly DefaultAnalyzerAssemblyLoader _loader;
+
+            public DirectoryLoadContext(string directory, DefaultAnalyzerAssemblyLoader loader)
+            {
+                _directory = directory;
+                _loader = loader;
+            }
+
+            protected override Assembly? Load(AssemblyName assemblyName)
+            {
+                // When we want to provide an analyzer with a possibly shared dependency, such as the compiler assemblies,
+                // we want to first search the assembly load context that this class is loaded into.
+                // This method of obtaining the context tries to account for the possibility that
+                // multiple versions of the compiler assemblies themselves could be hosted in different assembly load contexts in a single process.
+                var sharedContext = AssemblyLoadContext.GetLoadContext(typeof(DirectoryLoadContext).Assembly);
+
+                var alreadyLoadedAssembly = sharedContext?.Assemblies.FirstOrDefault(
+                    assembly => AssemblyName.ReferenceMatchesDefinition(assemblyName, assembly.GetName()));
+                if (alreadyLoadedAssembly is not null)
+                {
+                    return alreadyLoadedAssembly;
+                }
```

Comment:
So how exactly will the AllowList be constructed?

I was thinking we could try to get `System.Assembly` instances for MS.CA.CS and MS.CA.VB, call `GetReferencedAssemblies` on them, throw them in a set, and then check assembly names against it.

Alternatively, we could GetReferencedAssemblies on the Assembly this class is contained in, and also specially recognize the simple names for `MS.CA.CS` and `MS.CA.VB`.

Reply:
Or just hard code a list of assemblies the compiler itself depends on them then just have each analyzer / source generator use the already loaded ones that the compiler itself already loaded (as well as the dependencies of which) unless they specifically ask to load a newer version of said dependencies (excluding roslyn ones)

Unfortunate for them but only 1 version of roslyn must be used in the compiler at all times because that would be harder for the version of roslyn used by the compiler to comunicate with a different version of roslyn used by analyzers and source generators.

Because of that Microsoft plans to amend the roslyn docs to state that analyzers and source generators *should* never package *any* of roslyn's dlls in their nuget packages as that results in issues.

Reply:
> Or just hard code a list of assemblies the compiler itself depends on

This list is not static though. It changes, potentially, every time we update the .NET SDK and it's presently different between the various target frameworks that we use in the compiler. Yes the list could be static but then we'd most likely need to compensate by adding a test which dynamically calculated the reference list and validated it against the static one. Hence at that point may as well produce the dynamic one in the code itself.

Reply:
> I was thinking we could try to get System.Assembly instances for MS.CA.CS and MS.CA.VB, call GetReferencedAssemblies on them, throw them in a set, and then check assembly names against it.

That plus the referenced assemblies for MS.CA is how I was thinking of approaching this problem.

Reply:
Referenced assemblies for MS.CA.CS and MS.CA.VB will change as the MS.CA.CS and MS.CA.VB implementation evolves. I do not think you want to be in a place where accidental MS.CA.CS and MS.CA.VB implementation changes affect the set of platform assemblies usable by analyzers.

The runtime behavior needs to be in sync with the build time behavior. What is the set of reference assemblies that the analyzers reference by default at build time? Is it the whole netcoreapp or netstandard? The runtime behavior should match that.

Reply:
Usually one want to use System.Text.Json, the various roslyn dlls, YamlDotNet, System.Memory, etc.

But I would include the netstandard2.0 libraries as well too.

Reply:
>  What is the set of reference assemblies that the analyzers reference by default at build time? Is it the whole netcoreapp or netstandard? 

The recommendation is that the reference `netstandard2.0`. Due to build and packaging quirks though it's realistic that you will see several `System.*` DLLs in the analyzer NuPkg and as a result passed via `/analyzer`. If these references have the same simple name as references the compiler uses then we need to not load them into the analyzer ALC else there will be type mismatches with compiler APIs. 

> I do not think you want to be in a place where accidental MS.CA.CS and MS.CA.VB implementation changes affect the set of platform assemblies usable by analyzers.

The intent isn't to change which ones are used but rather which ALC loads them: the default or the analyzer ALC. 

The principal that we're operating under is that if the compiler is going to load a given DLL then it should never be loaded in the analyzer ALC. Doing the opposite is creating a likely type mismatch situation with our APIs.

Reply:
I guess another option is to check if the default ALC has a dll loaded do not try to load it into the analyzer / source generator ALC's and instead use the ones in the default ALC already. I think that is simple enough to do.

Reply:
> I guess another option is to check if the default ALC has a dll loaded do not try to load it into the analyzer / source generator ALC's and instead use the ones in the default ALC already. I think that is simple enough to do.

Concerns about this approach are outlined in https://github.com/dotnet/roslyn/pull/55098#discussion_r682607296.